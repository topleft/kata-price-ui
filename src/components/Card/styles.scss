@import '../../styles/variables.scss';

.card {
  align-items: center;
  background-color: $white;
  box-shadow: 0px 0px 20px $grey-light;
  display: flex;
  height: 400px;
  justify-content: space-between;
  padding: $space*2 $space*1.5 $space*4 $space*1.5;
  position: relative;
  flex-direction: column;
  width: 270px;


  @media screen and (min-width: 540px) {
  }

  &__highlight {
    color: $white;
    padding: $space/4 $space/1.5 $space/8 $space/1.5;
    background-color: $orange;
    text-transform: uppercase;
    font-size: $font-size-11;
    display: inline-block;

    border-radius: 2px;
  }

  &__{
    font-size: $font-size-11;
  }

  &__price {

    color: $black;

    &__currency-symbol {
      font-size: $font-size-16;
    }

    &__amount {
      font-size: $font-size-45;
      margin: 0 $space/3;
    }

    &__frequency {
      color: $grey-dark;
    }
  }

  &__list {
    align-self: flex-start;
    text-align: left;
    font-size: $font-size-14;
    color: $grey-dark;
    list-style-image: url('../../static/check-mark-button.svg');
    margin: 0 $space*2;

    &__item {
      margin: $space/2 0;
    }
  }

  &__button {
    border: 1px solid $grey-dark;
    border-radius: 4px;
    color: $grey-dark;
    padding: $space/1.5 $space*1.5;
    text-transform: uppercase;
    width: 150px;
  }

}
