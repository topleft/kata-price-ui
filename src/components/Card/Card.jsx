import React from 'react';
import PropTypes from 'prop-types';
import './styles.scss';

class Card extends React.PureComponent {
  constructor() {
    super();
  }

  renderHightlight() {
    const {highlightText} = this.props;
    const className = highlightText ? 'card__highlight' : 'card__highlight--no-background';
    return <div className={className}>{highlightText || '&nbsp;'}</div>
  }

  render() {
    const {
      buttonText,
      frequency,
      handleButtonClick,
      id,
      list,
      price,
      title,
    } = this.props;

    return (
      <div className='card'>
        {this.renderHightlight()}
        <div className='card__title'>{title}</div>
        <div className='card__price'>
          <span className='card__price__currency-symbol'>$</span>
          <span className='card__price__amount'>{price}</span>
          <span className='card__price__frequency'>/{frequency}</span>
        </div>
        <ul className='card__list'>
          {list.map((item, i) => <li key={i} className='card__list__item'>{item}</li>)}
        </ul>
        <button className='card__button' onClick={() => handleButtonClick(id)}>{buttonText}</button>
      </div>
    );
  }
}

Card.defaultProps = {
  list: [],
};

Card.propTypes = {
  buttonText: PropTypes.string,
  frequency: PropTypes.oneOf('mo', 'year'),
  handleButtonCLick: PropTypes.func,
  highlightText: PropTypes.string,
  id: PropTypes.string,
  list: PropTypes.arrayOf(PropTypes.string),
  price: PropTypes.number,
  title: PropTypes.string,
};

export default Card;
