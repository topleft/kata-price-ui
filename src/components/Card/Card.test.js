import React from 'react';
import {shallow} from 'enzyme';
import Card from './Card';

describe('PriceCard', () => {

  it('should render', () => {
    const wrapper = shallow(<Card/>);
    expect(wrapper.exists());
  });



  it('should render a highlight if highlightText prop is passed', () => {
    const highlightText = 'Sample';
    const wrapper = shallow(<Card highlightText={highlightText} />);
    expect(wrapper.find('.card__highlight').text()).toBe(highlightText);
  });

  it('should NOT render a highlight, but should render a space holder, if highlightText prop is falsey', () => {
    const wrapper = shallow(<Card />);

    wrapper.setProps({highlightText: ''});
    expect(wrapper.find('.card__highlight').length).toBe(0);
    expect(wrapper.find('.card__highlight--place-holder').length).toBe(1);
    expect(wrapper.find('.card__highlight--place-holder').text()).toBe('');

    wrapper.setProps({highlightText: undefined});
    expect(wrapper.find('.card__highlight').length).toBe(0);
    expect(wrapper.find('.card__highlight--place-holder').length).toBe(1);
    expect(wrapper.find('.card__highlight--place-holder').text()).toBe('');

    wrapper.setProps({highlightText: null});
    expect(wrapper.find('.card__highlight').length).toBe(0);
    expect(wrapper.find('.card__highlight--place-holder').length).toBe(1);
    expect(wrapper.find('.card__highlight--place-holder').text()).toBe('');
  });

  it('should render a title', () => {
    const title = 'Sample';
    const wrapper = shallow(<Card title={title} />);
    expect(wrapper.find('.card__title').text()).toBe(title);
  });

  it('should render price', () => {
    const wrapper = shallow(<Card price={100}/>);
    expect(wrapper.find('.card__price__amount').text()).toBe('100');
  });

  it('should render a frequency with a "/" appended', () => {
    const wrapper = shallow(<Card frequency={'mo'}/>);
    expect(wrapper.find('.card__price__frequency').text()).toBe('/mo');
  });

  it('should render list items', () => {
    const list = ['Information 1', 'Information 2', 'Information 3'];
    const wrapper = shallow(<Card list={list} />);
    const renderedList = wrapper.find('li.card__list__item');
    expect(renderedList.length).toBe(3);
    expect(renderedList.first().text()).toBe(list[0]);
  });

  it('should default the list prop to an empty array', () => {
    const wrapper = shallow(<Card/>);
    expect(wrapper.instance().props.list).toBeInstanceOf(Array);
    expect(wrapper.instance().props.list.length).toBe(0);
  });

  it('should render a button with passed in buttonText', () => {
    const wrapper = shallow(<Card buttonText={'Click Here'}/>);
    expect(wrapper.find('button.card__button').exists());
    expect(wrapper.find('button.card__button').text()).toBe('Click Here');
  })

  it('should call the handleButtonClick prop when button is clicked', () => {
    const clickFunc = jest.fn();
    const id = '1';
    const wrapper = shallow(<Card handleButtonClick={clickFunc} id={id}/>);
    const button = wrapper.find('button');
    expect(clickFunc).toHaveBeenCalledTimes(0);
    button.simulate('click');
    expect(clickFunc).toHaveBeenCalledTimes(1);
    expect(clickFunc).toHaveBeenCalledWith(id);
  });

  it('matches snapshot', () => {
    const card = {
      buttonText: 'Click Me',
      frequency: 'mo',
      handleButtonClick: () => console.log(),
      highlightText: 'Good Deal',
      id: '1',
      list: ['Cras sodales lobortis erat', 'Vitae pellentesque diam', 'Consequat eted tempus'],
      price: 100,
      title: 'Professional',
    }
    const tree = shallow(<Card {...card}/>);
    expect(tree).toMatchSnapshot();
  });

});
