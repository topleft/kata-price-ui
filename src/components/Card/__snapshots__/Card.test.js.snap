// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceCard matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Card
    buttonText="Click Me"
    frequency="mo"
    handleClick={[Function]}
    highlightText="Good Deal"
    id="1"
    list={
      Array [
        "Cras sodales lobortis erat",
        "Vitae pellentesque diam",
        "Consequat eted tempus",
      ]
    }
    price={100}
    title="Professional"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="card__highlight"
        >
          Good Deal
        </div>,
        <div
          className="card__title"
        >
          Professional
        </div>,
        <div
          className="card__price"
        >
          <span
            className="card__price__currency-symbol"
          >
            $
          </span>
          <span
            className="card__price__amount"
          >
            100
          </span>
          <span
            className="card__price__frequency"
          >
            /
            mo
          </span>
        </div>,
        <ul
          className="card__list"
        >
          <li
            className="card__list__item"
          >
            Cras sodales lobortis erat
          </li>
          <li
            className="card__list__item"
          >
            Vitae pellentesque diam
          </li>
          <li
            className="card__list__item"
          >
            Consequat eted tempus
          </li>
        </ul>,
        <button
          className="card__button"
          onClick={[Function]}
        >
          Click Me
        </button>,
      ],
      "className": "card",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Good Deal",
          "className": "card__highlight",
        },
        "ref": null,
        "rendered": "Good Deal",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Professional",
          "className": "card__title",
        },
        "ref": null,
        "rendered": "Professional",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              className="card__price__currency-symbol"
            >
              $
            </span>,
            <span
              className="card__price__amount"
            >
              100
            </span>,
            <span
              className="card__price__frequency"
            >
              /
              mo
            </span>,
          ],
          "className": "card__price",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "$",
              "className": "card__price__currency-symbol",
            },
            "ref": null,
            "rendered": "$",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": 100,
              "className": "card__price__amount",
            },
            "ref": null,
            "rendered": 100,
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "/",
                "mo",
              ],
              "className": "card__price__frequency",
            },
            "ref": null,
            "rendered": Array [
              "/",
              "mo",
            ],
            "type": "span",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <li
              className="card__list__item"
            >
              Cras sodales lobortis erat
            </li>,
            <li
              className="card__list__item"
            >
              Vitae pellentesque diam
            </li>,
            <li
              className="card__list__item"
            >
              Consequat eted tempus
            </li>,
          ],
          "className": "card__list",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "0",
            "nodeType": "host",
            "props": Object {
              "children": "Cras sodales lobortis erat",
              "className": "card__list__item",
            },
            "ref": null,
            "rendered": "Cras sodales lobortis erat",
            "type": "li",
          },
          Object {
            "instance": null,
            "key": "1",
            "nodeType": "host",
            "props": Object {
              "children": "Vitae pellentesque diam",
              "className": "card__list__item",
            },
            "ref": null,
            "rendered": "Vitae pellentesque diam",
            "type": "li",
          },
          Object {
            "instance": null,
            "key": "2",
            "nodeType": "host",
            "props": Object {
              "children": "Consequat eted tempus",
              "className": "card__list__item",
            },
            "ref": null,
            "rendered": "Consequat eted tempus",
            "type": "li",
          },
        ],
        "type": "ul",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Click Me",
          "className": "card__button",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Click Me",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="card__highlight"
          >
            Good Deal
          </div>,
          <div
            className="card__title"
          >
            Professional
          </div>,
          <div
            className="card__price"
          >
            <span
              className="card__price__currency-symbol"
            >
              $
            </span>
            <span
              className="card__price__amount"
            >
              100
            </span>
            <span
              className="card__price__frequency"
            >
              /
              mo
            </span>
          </div>,
          <ul
            className="card__list"
          >
            <li
              className="card__list__item"
            >
              Cras sodales lobortis erat
            </li>
            <li
              className="card__list__item"
            >
              Vitae pellentesque diam
            </li>
            <li
              className="card__list__item"
            >
              Consequat eted tempus
            </li>
          </ul>,
          <button
            className="card__button"
            onClick={[Function]}
          >
            Click Me
          </button>,
        ],
        "className": "card",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Good Deal",
            "className": "card__highlight",
          },
          "ref": null,
          "rendered": "Good Deal",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Professional",
            "className": "card__title",
          },
          "ref": null,
          "rendered": "Professional",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                className="card__price__currency-symbol"
              >
                $
              </span>,
              <span
                className="card__price__amount"
              >
                100
              </span>,
              <span
                className="card__price__frequency"
              >
                /
                mo
              </span>,
            ],
            "className": "card__price",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "$",
                "className": "card__price__currency-symbol",
              },
              "ref": null,
              "rendered": "$",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": 100,
                "className": "card__price__amount",
              },
              "ref": null,
              "rendered": 100,
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "/",
                  "mo",
                ],
                "className": "card__price__frequency",
              },
              "ref": null,
              "rendered": Array [
                "/",
                "mo",
              ],
              "type": "span",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <li
                className="card__list__item"
              >
                Cras sodales lobortis erat
              </li>,
              <li
                className="card__list__item"
              >
                Vitae pellentesque diam
              </li>,
              <li
                className="card__list__item"
              >
                Consequat eted tempus
              </li>,
            ],
            "className": "card__list",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "0",
              "nodeType": "host",
              "props": Object {
                "children": "Cras sodales lobortis erat",
                "className": "card__list__item",
              },
              "ref": null,
              "rendered": "Cras sodales lobortis erat",
              "type": "li",
            },
            Object {
              "instance": null,
              "key": "1",
              "nodeType": "host",
              "props": Object {
                "children": "Vitae pellentesque diam",
                "className": "card__list__item",
              },
              "ref": null,
              "rendered": "Vitae pellentesque diam",
              "type": "li",
            },
            Object {
              "instance": null,
              "key": "2",
              "nodeType": "host",
              "props": Object {
                "children": "Consequat eted tempus",
                "className": "card__list__item",
              },
              "ref": null,
              "rendered": "Consequat eted tempus",
              "type": "li",
            },
          ],
          "type": "ul",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Click Me",
            "className": "card__button",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Click Me",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
